import sys
import os
import base64
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import grpc
import medical_ai_pb2 as pb2
import medical_ai_pb2_grpc as pb2_grpc

def get_fake_request_data(stream_mode: bool, patient_dept: str):
    """ä½¿ç”¨å‡çš„æ‚£è€…æ•°æ®å’Œå›¾ç‰‡è¿›è¡Œæµ‹è¯•"""
    # å‡æ‚£è€…æ–‡æœ¬æ•°æ®
    patient_text_data = """
    **æ‚£è€…åŸºæœ¬ä¿¡æ¯**
    å§“åï¼šæµ‹è¯•æ‚£è€…
    æ€§åˆ«ï¼šå¥³
    å¹´é¾„ï¼š25 å²
    èŒä¸šï¼šæµ‹è¯•å‘˜
    èº«é«˜ï¼š165 cm
    ä½“é‡ï¼š60 kg
    BMIæŒ‡æ•°ï¼šçº¦ 22.0
    è¿‡æ•å²ï¼šæ— 
    æ—¢å¾€ç—…å²ï¼šæ— ç‰¹æ®Šè¯´æ˜

    **ä¸»è¯‰ä¸ç°ç—…å²**
    ä¸»è¯‰ï¼šæµ‹è¯•ç—‡çŠ¶ã€‚
    ç—‡çŠ¶æè¿°ï¼šè¿™æ˜¯ä¸€ä¸ªæµ‹è¯•æ¡ˆä¾‹ï¼Œç”¨äºéªŒè¯ç³»ç»ŸåŠŸèƒ½ã€‚
    """
    
    # ä½¿ç”¨å‡çš„Base64å›¾ç‰‡æ•°æ®ï¼ˆä¸€ä¸ªå°çš„é€æ˜PNGï¼‰
    fake_image_base64 = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg=="
    print(f"âœ… ä½¿ç”¨å‡å›¾ç‰‡æ•°æ® (Base64é•¿åº¦: {len(fake_image_base64)} å­—ç¬¦)")
    
    return pb2.AnalysisRequest(
        patient_text_data=patient_text_data,
        image_base64=fake_image_base64,
        stream=stream_mode,
        patient_department=patient_dept
    )

def sync_call(patient_dept: str):
    print("\n" + "="*50)
    print(f"ã€åŒæ­¥è°ƒç”¨æµ‹è¯•ã€‘ç—…äººé€‰æ‹©ç§‘å®¤ï¼š{patient_dept}")
    print("="*50)
    
    with grpc.insecure_channel('127.0.0.1:50051') as channel:
        stub = pb2_grpc.MedicalAIServiceStub(channel)
        request = get_fake_request_data(stream_mode=False, patient_dept=patient_dept)
        response_iterator = stub.ProcessMedicalAnalysis(request)
        
        for response_chunk in response_iterator:
            if response_chunk.is_end:
                # åŒæ­¥ï¼šbytesâ†’Protobufå¯¹è±¡ï¼ˆååºåˆ—åŒ–ï¼‰
                sync_report = pb2.AnalysisReport.FromString(response_chunk.chunk_data)
                if sync_report.status == "SUCCESS":
                    # æŠ¥å‘Šæ˜¯UTF-8å­—ç¬¦ä¸²ï¼Œç›´æ¥æ‰“å°
                    print("âœ… åŒæ­¥è°ƒç”¨æˆåŠŸï¼Œå®Œæ•´æŠ¥å‘Šï¼š")
                    print(sync_report.structured_report)
                else:
                    print(f"âŒ åŒæ­¥è°ƒç”¨å¤±è´¥")
                    print(f"   çŠ¶æ€: {sync_report.status}")
                    print(f"   æ¶ˆæ¯: {sync_report.message}")
                    if sync_report.structured_report:
                        print(f"   è¯¦æƒ…: {sync_report.structured_report}")
                break

def stream_call(patient_dept: str):
    print("\n" + "="*50)
    print(f"ã€æµå¼è°ƒç”¨æµ‹è¯•ã€‘ç—…äººé€‰æ‹©ç§‘å®¤ï¼š{patient_dept}")
    print("="*50)
    
    with grpc.insecure_channel('127.0.0.1:50051') as channel:
        stub = pb2_grpc.MedicalAIServiceStub(channel)
        request = get_fake_request_data(stream_mode=True, patient_dept=patient_dept)
        response_iterator = stub.ProcessMedicalAnalysis(request)
        
        print("ğŸ“¥ å¼€å§‹æ¥æ”¶æµå¼æ•°æ®ï¼š")
        for response_chunk in response_iterator:
            # æµå¼ï¼šbytesâ†’UTF-8å­—ç¬¦ä¸²
            chunk_str = response_chunk.chunk_data.decode('utf-8')
            if chunk_str == "[STREAM_END]":
                print("\nğŸ“¥ æµå¼æ¥æ”¶å®Œæ¯•ï¼ˆæ”¶åˆ°ç»“æŸæ ‡è®°ï¼‰")
                break
            print(f"[æµå¼chunk] {chunk_str}", end="")
            if not response_chunk.is_end:
                print("\n" + "-"*30)

if __name__ == "__main__":
    print("\n" + "="*80)
    print("ğŸŒŸ åŒ»ç–—AIæœåŠ¡ gRPC å®¢æˆ·ç«¯æµ‹è¯•ï¼ˆä½¿ç”¨å‡æ•°æ®ï¼‰")
    print("="*80)
    
    # æµ‹è¯• 1: åŒæ­¥è°ƒç”¨ + æ­£ç¡®ç§‘å®¤
    sync_call(patient_dept="è€³é¼»å–‰ç§‘")
    
    # æµ‹è¯• 2: æµå¼è°ƒç”¨ + æ­£ç¡®ç§‘å®¤
    # stream_call(patient_dept="è€³é¼»å–‰ç§‘")
    
    # æµ‹è¯• 3: åŒæ­¥è°ƒç”¨ + é”™è¯¯ç§‘å®¤
    # sync_call(patient_dept="å†…ç§‘")
    
    # æµ‹è¯• 4: æµå¼è°ƒç”¨ + é”™è¯¯ç§‘å®¤
    # stream_call(patient_dept="å†…ç§‘")
    
    print("\n" + "="*80)
    print("âœ… æµ‹è¯•å®Œæˆ")
    print("="*80)